src/connections.rs

```diff
@@ -225,8 +230,17 @@ impl IncomingMessages {
                 error!("Failed to read incoming message on bi-stream: {}", err);
                 None
             }
-            Some(Ok((send, mut recv))) => match read_bytes(&mut recv).await {
-                Ok(bytes) => Some((bytes, send, recv)),
+            Some(Ok((mut send, mut recv))) => match read_bytes(&mut recv).await {
+                Ok(WireMsg::UserMsg(bytes)) => Some((bytes, send, recv)),
+                Ok(WireMsg::EndpointEchoReq) => {
+                    let message = WireMsg::EndpointEchoResp(peer_addr);
+                    message.write_to_stream(&mut send).await.ok()?;
+                    None
```

Comment:
I think we shouldn't return `None` in this case as that would mean for the user that there won't be no more messages arriving in the stream, and therefore the user will end the loop. We should send the echo response, but then go back and read the next message by continuing the loop, so if next message is for the user it will receive it as thee next message without noticing there was an echo response sent in beforehand.